### PR TITLE
feat: added `mdxSource` when the `<Canvas>` not wrapping any `<Story>`.

### PR DESCRIPTION
**Pull request by zhyd1997**
Issue: https://github.com/storybookjs/storybook/issues/18342

## What Changed

generate`mdxSource` when `<Canvas>` has no `<Story>` child.

## How to test

[mdx1-csf](https://github.com/storybookjs/mdx1-csf) works for this situation because it added `mdxSource` manually:

https://github.com/storybookjs/mdx1-csf/blob/4b35b053979b71fd038370566520f1726939505f/src/sb-mdx-plugin.ts#L432-L448

and `mdx2-csf` (the package is _in progress_) has no `mdxSource` injected.

so we just need to compare the `mdxSource`, if `true`, then it passes.

### Test:

1. wrap the 2 stories (**Note**: remove all comments) in test files: 
  - [the 1st one](https://storybook.js.org/docs/react/writing-docs/doc-block-canvas#rendering-multiple-stories) (`<Canvas>` contains `<Story>` children and more than one)

  - [ the 2nd one](https://storybook.js.org/docs/react/writing-docs/doc-block-canvas#non-story-content) (`<Canvas> ` has no `<Story>` children but more than one other elements or React components)

2. log the `mdxSource` (`console.warn(getMdxSource(...))`) below this line:
  - mdx1-csf: https://github.com/storybookjs/mdx1-csf/blob/4b35b053979b71fd038370566520f1726939505f/src/sb-mdx-plugin.ts#L438
  - mdx2-csf: https://github.com/zhyd1997/mdx2-csf/blob/0b74cabbec09f87e7915c69e68514c5cf186ade8/src/mdx2.ts#L48

3. run `yarn test` and compare that if they are the same string or `decodeURI(mdxSource)` (expect: the are the same value).

## Change Type

- [ ] `maintenance`
- [ ] `documentation`
- [ ] `patch`
- [x] `minor`
- [ ] `major`

### CODE REVIEW

**Comment by shilman** [comment]
@zhyd1997 thank you for this PR. can you please add a test to `[mdx2.test.ts](https://github.com/storybookjs/mdx2-csf/blob/main/src/mdx2.test.ts)` to test the feature?

**Comment by zhyd1997** [comment]
Hi @shilman 

I tried for a while but can not find a way to test what `mdxSource` will be returned, which will be used as one of [Canvas](https://github.com/storybookjs/storybook/blob/c503691a1d42e0c00b7e14f706b149b79b71de14/code/lib/blocks/src/blocks/Canvas.tsx#L18) component props.

File: src/mdx2.test.ts [hunk 1]
```diff
@@ -125,6 +167,85 @@ describe('mdx2', () => {
...
+    expect(getCanvasWithoutStoryElements(canvasElements).length).toBe(1);
+    // test `getMdxSource` function.
+    const canvasWithNoStoryContainers = getCanvasWithoutStoryElements(canvasElements);
+    const encodedMdxSource = getMdxSource(canvasWithNoStoryContainers[0].children);
+    expect(decodeURI(encodedMdxSource)).toMatchInlineSnapshot(
+      dedent`
+        <h2>Some here</h2>
+        <MyComponent />
+      `
+    );
```
Comment (zhyd1997) [inline src/mdx2.test.ts:247]: the snapshot doesn't match:

```diff
- <h2>Some here</h2>
+ <h2>{"Some here"}</h2>
```
there're extra brace pair and double quotes wrapping `expression` value.

Comment (zhyd1997) [reply src/mdx2.test.ts]: @shilman 

I confirmed this issue will occur even if the `<Canvas>` wraps `<Story>`:

Screen Shot 2022-08-18 at 23 42 22

```tsx
import { Canvas, Meta, Story } from '@storybook/addon-docs';

import { MyComponent } from './MyComponent';

<Meta title="test" />

<Canvas>
  <Story name="story">
  <h2>Some here</h2>
  </Story>
  <MyComponent />
</Canvas>
```

so IMO this is another issue, and the test case will **fail** as expected, we need a separate PR to fix it.

**Comment by zhyd1997** [comment]
Screen Shot 2022-08-19 at 00 08 37

File: src/mdx2.test.ts (hunk 1 above)
Comment (zhyd1997) [reply src/mdx2.test.ts]: It's valid `tsx`, not sure for other framework (Angular, [Svelte](https://svelte.dev/)).

**Comment by shilman** [comment]
Thank you @zhyd1997 !!! Closing this in favor of #25 which builds on your changes

**Pull request closed without merging**